### ISSUE TITLE
Vectorizing interleaved RGBA/RGB images is too hard

### ISSUE THREAD

**Issue by steven-johnson** [halide/Halide#6630]
When interfacing with existing systems (e.g., OpenCV), it's generally critical to be able to write efficient code for interleaved (not planar) RGB and RGBA images; while this is possible, it's fraught with error and getting things Just Right -- e.g., you can have an approach that looks plausible but will end up scalarizing or performing poorly. I don't have a solution, but want to open discussion as to whether there are (1) outright bugs to be fixed, (2) scheduling or internal improvements that could make things easier or more reliable, or (3) something else.

**Comment by steven-johnson** [comment]
Here's today's example: let's say I need to process dense interleaved RGBA data, in the (x, y, c) order that Buffer::make_interleaved() produces. In this case, I need to examine just the RGB components (ignoring A), and reduce to a "luminance" or similar. Here was my first attempt:

    {
        Var x, y, c;
        constexpr int input_channels = 4;
        constexpr int edge_len = 128;

        ImageParam rgba(UInt(8), 3);

        // interleaved RGBA, in (x, y, c) index order.
        // We'll set explicit extents to simplify the generated code.
        rgba.dim(0).set_min(0).set_extent(edge_len).set_stride(input_channels);
        rgba.dim(1).set_min(0).set_extent(edge_len).set_stride(Expr());
        rgba.dim(2).set_min(0).set_extent(input_channels).set_stride(1);

        // Wrapper function around the input, for scheduling
        Func pixel;
        pixel(x, y, c) = rgba(x, y, c);

        // Calculate the "luma" from max(R,G,B), ignoring alpha
        Func luma;
        luma(x, y) = Halide::max(pixel(x, y, 0), pixel(x, y, 1), pixel(x, y, 2));

        Var cx("cx"), cxo("cxo"), cxi("cxi"), xo("xo"), xi("xi");

        const int vec_len = 16;  // u8x16 for 128-bit simd
        pixel.store_in(MemoryType::Register)
            .bound(c, 0, input_channels)
            .fuse(c, x, cx)
            .split(cx, cxo, cxi, vec_len)
            .vectorize(cxi)
            .compute_at(luma, xo);

        luma.store_in(MemoryType::Stack)
            .split(x, xo, xi, vec_len/4)
            .vectorize(xi);

        // Set bounds on luma too, to simplify the generated code
        luma.bound(x, 0, edge_len).bound(y, 0, edge_len);
        luma.compile_jit();
    }

This produces an inner loop like this:

    allocate pixel[uint8 * 16] in Register
    produce pixel {
     pixel[interleave_vectors(ramp(0, 1, 4), ramp(4, 1, 4), ramp(8, 1, 4), ramp(12, 1, 4))] = rgba[ramp((luma.s0.x.xo*16) + t11, 1, 16)]
    }
    consume pixel {
     luma[ramp((luma.s0.x.xo*4) + t12, 1, 4)] = max(pixel[ramp(0, 1, 4)], max(pixel[ramp(4, 1, 4)], pixel[ramp(8, 1, 4)]))
    }

Note that the pixel calculation involves `interleave_vectors` -- definitely suboptimal. Interestingly, I can "solve" this by changing the addressing convention on the input buffer; the memory layout is the same, but by indexing with (c, x, y), instead of the more conventional (x, y, c), I get a better result:

    {
        Var x, y, c;
        constexpr int input_channels = 4;
        constexpr int edge_len = 128;

        ImageParam rgba(UInt(8), 3);

        // interleaved RGBA, but in (c, x, y) index order.
        // We'll set explicit extents to simplify the generated code.
        rgba.dim(0).set_min(0).set_extent(input_channels).set_stride(1);
        rgba.dim(1).set_min(0).set_extent(edge_len).set_stride(input_channels);
        rgba.dim(2).set_min(0).set_extent(edge_len).set_stride(Expr());

        // Wrapper function around the input, for scheduling
        Func pixel;
        pixel(c, x, y) = rgba(c, x, y);

        // Calculate the "luma" from max(R,G,B), ignoring alpha
        Func luma;
        luma(x, y) = Halide::max(pixel(0, x, y), pixel(1, x, y), pixel(2, x, y));

        Var cx("cx"), cxo("cxo"), cxi("cxi"), xo("xo"), xi("xi");

        const int vec_len = 16;  // u8x16 for 128-bit simd
        pixel.store_in(MemoryType::Register)
            .bound(c, 0, input_channels)
            .fuse(c, x, cx)
            .split(cx, cxo, cxi, vec_len)
            .vectorize(cxi)
            .compute_at(luma, xo);

        luma.store_in(MemoryType::Stack)
            .split(x, xo, xi, vec_len/4)
            .vectorize(xi);

        // Set bounds on luma too, to simplify the generated code
        luma.bound(x, 0, edge_len).bound(y, 0, edge_len);
        luma.compile_jit();
    }


With the inner loop like:

    allocate pixel[uint8 * 16] in Register
    produce pixel {
     pixel[ramp(0, 1, 16)] = rgba[ramp((luma.s0.x.xo*16) + t3, 1, 16)]
    }
    consume pixel {
     luma[ramp((luma.s0.x.xo*4) + t4, 1, 4)] = max(pixel[ramp(0, 4, 4)], max(pixel[ramp(1, 4, 4)], pixel[ramp(2, 4, 4)]))
    }

This is definitely a more appealing result, but it shouldn't be necessary to rearrange the dimensions to achieve this. (Adding .reorder() or .reorder_storage() to the first case don't appear to make any difference.)


So is this a flat-out bug and it should work? Or is there additional scheduling I need to do in the first case to make it work like the second?

**Comment by abadams** [comment]
The thing that's making a mess in the first case is you're fusing an outer var with an inner var. The current order of the vars matters in fuse()

These days I think it usually works better to vectorize multiple vars than to fuse two vars and vectorize the fused result.

**Comment by steven-johnson** [comment]
> These days I think it usually works better to vectorize multiple vars than to fuse two vars and vectorize the fused result.

But will that work reasonably well when one of the vars is of extent=3?

**Comment by abadams** [comment]
Yeah, you just get a 96-element vector or whatever. It works out as OK as anything else does for this problem.

**Comment by steven-johnson** [comment]
> The thing that's making a mess in the first case is you're fusing an outer var with an inner var. The current order of the vars matters in fuse()

For future reference: in the example I gave, adding `.reorder(c, x, y).reorder_storage(c, x, y)` allowed me to to get the code generation I wanted.

There's still a regrettable drawback here that I haven't figured out: If I want this code to allow for RGB (as well as RGBA) inputs, I can't just use `.specialize()`, unfortunately: the `.bound(c, 0, input_channels)` directive for `pixel` is required to get the registerization sane, and `.bound()` can't be controlled via `.specialize()`. (I can of course work around it by compiling the entire function twice, once for each value, but that produces nontrivial code duplication. Would be nice if there was a better option.)

**Comment by abadams** [comment]
.bound takes Exprs too, I believe. Did you try the following?

.bound(c, 0, input.dim(2).extent());

.specialize(input.dim(2).extent() == 3);
.specialize(input.dim(2).extent() == 4);

**Comment by steven-johnson** [comment]
Aha: moving the .specialize() calls higher up (e.g. to `luma`) does the trick. Awesome!

**Comment by zvookin** [comment]
All I can think of to help with this sort of thing is having routines that encapsulate scheduling patterns. Little recipes for scheduling that are parameterized enough to be broadly useful but not so much as to be as difficult as scheduling itself in real use. This would either be a new header, or more than one, of scheduling utilities, or possibly just in Halide.h.

I'm not a huge fan of hiding scheduling code in application routines, but providing a standard set of encapsulations seems useful for cases where there is a clearly identified and widely applicable pattern of reuse.
